pyomo/core/expr/base.py

```diff
@@ -0,0 +1,387 @@
+
+        This method contrasts with the is_fixed() method.  This method
+        returns True if the expression is an atomic constant, that is it
+        is composed exclusively of constants and immutable parameters.
+        NumericValue objects returning is_constant() == True may be
+        simplified to their numeric value at any point without warning.
+
+        Note:  This defaults to False, but gets redefined in sub-classes.
+        """
+        return False
+
+    def is_fixed(self):
+        """
+        Return :const:`True` if this expression contains no free variables.
+
+        Returns:
+            A boolean.
+        """
+        return _expression_is_fixed(self)
+
+    def _is_fixed(self, values):
+        """
+        Compute whether this expression is fixed given
+        the fixed values of its children.
+
+        This method is called by the :class:`_IsFixedVisitor
+        <pyomo.core.expr.current._IsFixedVisitor>` class.  It can
+        be over-written by expression classes to customize this
+        logic.
+
+        Args:
+            values (list): A list of boolean values that indicate whether
+                the children of this expression are fixed
+
+        Returns:
+            A boolean that is :const:`True` if the fixed values of the
+            children are all :const:`True`.
+        """
+        return all(values)
+
+    def is_potentially_variable(self):
+        """
+        Return :const:`True` if this expression might represent
+        a variable expression.
+
+        This method returns :const:`True` when (a) the expression
+        tree contains one or more variables, or (b) the expression
+        tree contains a named expression. In both cases, the
+        expression cannot be treated as constant since (a) the variables
+        may not be fixed, or (b) the named expressions may be changed
+        at a later time to include non-fixed variables.
+
+        Returns:
+            A boolean.  Defaults to :const:`True` for expressions.
+        """
+        return True
+
+    def is_named_expression_type(self):
+        """
+        Return :const:`True` if this object is a named expression.
+
+        This method returns :const:`False` for this class, and it
+        is included in other classes within Pyomo that are not named
+        expressions, which allows for a check for named expressions
+        without evaluating the class type.
+
+        Returns:
+            A boolean.
+        """
+        return False
+
+    def is_expression_type(self, expression_system=None):
+        """
+        Return :const:`True` if this object is an expression.
+
+        This method obviously returns :const:`True` for this class, but it
+        is included in other classes within Pyomo that are not expressions,
+        which allows for a check for expressions without
+        evaluating the class type.
+
+        Returns:
+            A boolean.
+        """
+        return expression_system is None \
```

Comment:
Ahh.  `m.e.is_expression_type()` (i.e., when `expression_system=None`) returns true when the object belongs to *any* expression system.  Otherwise, it only returns True when the object is a member of the *specified* expression system.